PeriodogramAnalysis/test_power_density_statistics.py

```diff
@@ -0,0 +1,237 @@
+    result = analyzer(mock_signal_generator())
+
+    assert mock_compute_psd.call_count == 2
+    assert result == defaultdict(dict)
+
+    # Test __call__ of children class
+    analysis_wel = SpectrumAnalysisWelch()
+    psd_dict_wel = analysis_wel(mock_signal_generator())
+    analysis_per = SpectrumAnalysisPeriodogram()
+    psd_dict_per = analysis_per(mock_signal_generator())
+
+    psd_dict_list = [psd_dict_wel, psd_dict_per]
+
+    for psd_dict in psd_dict_list:
+        assert isinstance(psd_dict, dict)
+        assert len(psd_dict) == 2
+
+        # check the content of spec_dict
+        for channel, channel_data in psd_dict.items():
+            assert "freqs" in channel_data
+            assert "psd" in channel_data
+
+
+def test_SpectrumAnalysisWelch() -> None:
+    """
+    Test if welch method is correct, for this input signal, the psd should have peaks at given frequencies,
+    So I test if the returned result has peaks there.
+    If it failed, it should be wrong with welch method, perhaps the calculation is wrong.
+    """
+    analysis = SpectrumAnalysisWelch()
+    psd_dict = analysis(mock_signal_generator())
+
+    for channel, channel_data in psd_dict.items():
+        for chunk in range(len(psd_dict[channel]["freqs"])):
+            freqs = channel_data["freqs"][chunk]
+            freqs = np.array(freqs)
+            psd = channel_data["psd"][chunk]
+            total_psd_sum = np.sum(psd)
+            expected_freqs = [5, 10, 15, 20]
+
+            range_psd_sum = 0
+            for exp_freq in expected_freqs:
+                range_psd_sum += np.sum(
+                    psd[(freqs >= exp_freq - 1) & (freqs <= exp_freq + 1)]
+                )
+
+                idx = np.argmin(np.abs(freqs - exp_freq))
+                # Test the peaks of psd is as expected
+                assert np.isclose(freqs[idx], exp_freq)
+            # Test the psd around peaks comprises more than 90% of total psd
+            assert range_psd_sum / total_psd_sum >= 0.9
+
+
+def test_SpectrumAnalysisPeriodogram() -> None:
+    """
+    Same as previous one.
+    """
+    analysis = SpectrumAnalysisPeriodogram()
+    psd_dict = analysis(mock_signal_generator())
+
+    for channel, channel_data in psd_dict.items():
+        for chunk in range(len(psd_dict[channel]["freqs"])):
+            freqs = channel_data["freqs"][chunk]
+            freqs = np.array(freqs)
+            psd = channel_data["psd"][chunk]
+            total_psd_sum = np.sum(psd)
+            expected_freqs = [5, 10, 15, 20]
+
+            range_psd_sum = 0
+            for exp_freq in expected_freqs:
+                range_psd_sum += np.sum(
+                    psd[(freqs >= exp_freq - 1) & (freqs <= exp_freq + 1)]
+                )
+
+                idx = np.argmin(np.abs(freqs - exp_freq))
+                assert np.isclose(freqs[idx], exp_freq)
+
+            assert range_psd_sum / total_psd_sum >= 0.9
+
+
+def test_SpectrumAnalysisWelch_compute_psd() -> None:
```

Comment:
This test would be not needed if you use handle the signal processing within the generator.

Reply:
Have deleted it now.